File: Dynamic Programming/Java/Longest Increasing Path in a Matrix.java
// Problem link: https://leetcode.com/problems/longest-increasing-path-in-a-matrix/
// Time: O(n * m)
// Space: O(n * m)
class Solution {
    public int longestIncreasingPath(int[][] matrix) {
        int rows = matrix.length, cols = matrix[0].length, result = 0;
        int[][] cache = new int[rows][cols];

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                result = Math.max(result, this.dfs(row, col, Integer.MIN_VALUE, matrix, cache));
            }
        }

        return result;
    }

    private int dfs(int row, int col, int prev, int[][] matrix, int[][] cache) {
        if(row < 0 || col < 0 || col == matrix[0].length || row == matrix.length || matrix[row][col] <= prev) {
            return 0;
        }
        else if(cache[row][col] != 0) {
            return cache[row][col];
        }

        int[][] directions = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
        int count = 0;

        for(int[] dir : directions) {
            count = Math.max(count, this.dfs(row + dir[0], col + dir[1], matrix[row][col], matrix, cache));
        }

        cache[row][col] = count + 1;
        return cache[row][col];
    }
}

File: Dynamic Programming/Python/Longest Increasing Path in a Matrix.py
# Problem link: https://leetcode.com/problems/longest-increasing-path-in-a-matrix/
# Time: O(n * m)
# space: O(n * m)
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        rows, cols, result = len(matrix), len(matrix[0]), 0
        cache = [[0] * cols for _ in range(rows)]

        for row i in range(rows):
            for col in range(cols):
                result = max(result, self.dfs(row, col, matrix, cache, -inf), result)
        
        return result
    
    def dfs(self, row, col, matrix, cache, prev):
        if row < 0 or col < 0 or row == len(matrix) or col == len(matrix[0]) or matrix[row][col] <= prev:
            return 0
        elif cache[row][col]:
            return cache[row][col]
        
        directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        for dr, dc in directions:
            count = 1 + max(count, self.dfs(row + dr, col + dc, matrix, cache, matrix[row][col]))

        cache[row][col] = count
        return count
